File: Heapq/src/main/java/Solution23.java
import common.ListNode;
import java.util.*;

public class Solution23 {

    public ListNode mergeKLists(ListNode[] lists) {

        class MyNode implements Comparable<MyNode> {
            int val;
            ListNode node;

            MyNode(int val, ListNode node) {
                this.val = val;
                this.node = node;
            }

            @Override
            public int compareTo(MyNode o) {
                return this.val - o.val;
            }

        }

        PriorityQueue<MyNode> queue = new PriorityQueue<MyNode>();

        for (ListNode listNode : lists) {
            if (listNode != null) {
                queue.offer(new MyNode(listNode.val, listNode));
            }
        }
        ListNode head = new ListNode(0);
        ListNode tail = head;

        while (!queue.isEmpty()) {
            MyNode min = queue.poll();

            tail.next = min.node;
            tail = tail.next;

            if (min.node.next != null) {
                queue.offer(new MyNode(min.node.next.val, min.node.next));
            }
        }
        return head.next;
    }
}

File: Heapq/src/main/python/solution23.py
from platform import node
from typing import List
import heapq


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:

        pq = []

        for list_node in lists:
            if list_node:
                heapq.heappush(pq, MyNode(list_node.val, list_node))

        print(pq)

        head = ListNode(0, None)
        tail = head

        while len(pq) != 0:
            min_top = heapq.heappop(pq)
            tail.next = min_top.node
            tail = tail.next

            if min_top.node.next:
                heapq.heappush(pq, MyNode(min_top.node.next.val, min_top.node.next))

        return head.next


class MyNode:
    def __init__(self, val=0, node=None):
        self.val = val
        self.node = node

    def __lt__(self, other):
        return self.val < other.val


s = Solution()
s.mergeKLists([[]])